### PR TITLE
Proxied remote gems should never be added as local gems

### PR DESCRIPTION
**Pull request by jeremy**
Since local gems take precedence over remote gems, if proxying a remote gem
adds it as a local gem, then that one proxied gem now takes precedence over
all other versions of the remote gem.

To fix, proxy just the .gem file. Don't add to the local gem store.

This is a problem for existing servers upgrading to 1.3.0, since existing proxied
remote gems are polluting the local gem store. The local proxied gems need to
be yanked so they no longer take precedence over remote gems. There is not
a clear path to doing that since, as far as I can tell, Geminabox does not track
any state to distinguish proxy-cached gems from other local gems.

This should also result in a significant performance boost, since installing
a proxied remote gem will no longer add a local gem and kick off a very
expensive reindex.

Note: There is no test coverage for this part of the server.

Follow-up to #339.

### CODE REVIEW

**Comment by pingvistyle** [comment]
I have same problem. I use 1.3.0 version.
If geminbox downloaded some gem wiith version N in firt time and I change version in Gemlock.file to N-1 or N+1 I get message:

```
Your bundle is locked to i18n (1.8.9), but that version could not be found in any of
the sources listed in your Gemfile. If you haven't changed sources, that means the
author of i18n (1.8.9) has removed it. You'll need to update your bundle to a version
other than i18n (1.8.9) that hasn't been removed in order to install.
```

**Review by simi — APPROVED** [review]

**Comment by sriedel** [comment]
Just to clarify the new intended behavior with this PR: gems from rubygems.org will no longer be saved by geminabox in a local cache at all, and will only serve gems previously deployed to geminabox with `gem geminabox mygem.gem` from local storage?

If so, this raises two issues:
- Having a local cache of rubygems.org gems is a huge plus, as deployments could be independent of rubygems being up or down. This should then probably be a separate feature request though.
- Some indication in the Changelog or README on how to clean up any previously cached gems from rubygems.org in geminaboxes local storage would be good for maintainers of geminabox instances that have been running for a while and will now have a mix of rubygems and their own gems stored, where the rubygems gems will no longer be updated to more current versions.

**Comment by jeremy** [comment]
@sriedel Ideally, geminabox acts as a caching proxy for remote gems files without adding the remote gems to the local index. This PR scope is limited to correcting the "remote gems leak into local gems" behavior, however.

I'd like a guide to how to clean up leaked gems, too! As far as I can tell, we can't distinguish local from remote gems without hunting through log files to identify gem pushes.

**Comment by sriedel** [comment]
In my case I have the luxury of knowing which gems should be local and which are cached from rubygems. So with my limited understand it may boil down to "removing non-local gems from the filesystem cache directory and rebuild the local index". 

The first part is easily done, if tedious. But how would the local index be rebuilt? Would pushing a new local gem version be the simplest way?

**Comment by jeremy** [comment]
That'd do it.

**Comment by NobodysNightmare** [comment]
Anything that can be done to bring this forward? This starts to be problematic for us and we are currently weighing our options:
* downgrade to 1.2
* wait for the next release
* use a custom build that's already including this PR

But maybe you have thoughts on this, e.g. whether you are still undecided if this is the correct solution to the problem or any other caveats one should be aware of before using this PR prematurely :)

**Comment by NobodysNightmare** [comment]
Just in case this is helpful to anyone: We are currently using a local fork of `geminabox` with this PR applied, so it can continue serving a mix of internal and public gems. *It is working fine for us.*

The greatest challenge was to remove all public gems from the local index.

This was our approach (also hoping it may be helpful to others):

1. Take a backup of the gems on disk
2. Delete all gems that we identified as public gems from disk
3. Force reindex through `https://gems.example.com/reindex`
4. Use it

In step 4 you might find errors (deleted too much/too few). If they occur either delete more files or restore missing files from backup and reindex again.

One additional caveat we discovered: In case you use gems that transitioned from internal to public or vice versa, you will have to decide from where to get them (i.e. either all from `geminabox` or all from rubygems.org). For us this only affected two or so gems, where we could ensure to only get them from rubygems.org after a gem update in our applications.

**Comment by ahorek** [comment]
after applying the patch I saw a data corruption, but the real problem is elsewhere, here's the fix https://github.com/geminabox/geminabox/pull/357

👍 for this change

**Comment by seandilda** [comment]
This feels like the wrong solution.  Or at least a solution that needs a config option.

I run an internal proxy using geminabox that intentionally caches the gems locally.  This is needed to fully support the
`Geminabox.allow_remote_failure = true` feature.

I believe the true fix may be to stop supporting the 'local gems + proxy' setup.  For my org I run two copies of geminabox.  The first allows uploading of internal gems and does not support proxying.   The second instance does not allow upload of gems, but does handle proxying of gems from rubygems.org   Having the split servers allows for cleanly separating gem sources in Gemfile.

For anyone looking for another solution to this, setting `Geminabox.rubygems_proxy_merge_strategy = :combine_local_and_remote_gem_versions` seems to be working well.

**Comment by zinga666** [comment]
Oh boy, @seandilda this

>For anyone looking for another solution to this, setting Geminabox.rubygems_proxy_merge_strategy = :combine_local_and_remote_gem_versions seems to be working well.

was exactly what I needed (but didn't know existed). Thanks a lot!

**Comment by simi** [comment]
This make it work again for my usecase as well. Thanks @seandilda for sharing that. Would it be possible to document this option at least in README?

**Comment by fauno** [comment]
Hi, we stumbled with this behavior too and started a different discussion at #370

**Comment by skaes** [comment]
> The greatest challenge was to remove all public gems from the local index.
> 
> This was our approach (also hoping it may be helpful to others):
> 
> 1. Take a backup of the gems on disk
> 2. Delete all gems that we identified as public gems from disk
> 3. Force reindex through `https://gems.example.com/reindex`
> 4. Use it

There is actually a slightly more sophisticated and automatic way of doing this:

For each local gem, check whether all local gem files for that gem are present on rubygems.org and the file content has the same SHA256. If so, you can delete all local versions of that gem for  your server. It's unlikely to be a locally developed gem in that case.

**Comment by skaes** [comment]
I believe all the topics here are now fully addressed in https://github.com/geminabox/geminabox/pull/435.

**Pull request closed without merging**
